// @aztec/accounts@0.28.1/dest/single_key/account_contract.js
import { generatePublicKey } from '@aztec/aztec.js';
import { AuthWitness } from '@aztec/circuit-types';
import { Schnorr } from '@aztec/circuits.js/barretenberg';
import { DefaultAccountContract } from '../defaults/account_contract.js';
import { SchnorrSingleKeyAccountContractArtifact } from './artifact.js';
/**
 * Account contract that authenticates transactions using Schnorr signatures verified against
 * the note encryption key, relying on a single private key for both encryption and authentication.
 */
export class SingleKeyAccountContract extends DefaultAccountContract {
    constructor(encryptionPrivateKey) {
        super(SchnorrSingleKeyAccountContractArtifact);
        this.encryptionPrivateKey = encryptionPrivateKey;
    }
    getDeploymentArgs() {
        return undefined;
    }
    getAuthWitnessProvider({ partialAddress }) {
        return new SingleKeyAuthWitnessProvider(this.encryptionPrivateKey, partialAddress);
    }
}
/**
 * Creates auth witnesses using Schnorr signatures and including the partial address and public key
 * in the witness, so verifiers do not need to store the public key and can instead validate it
 * by reconstructing the current address.
 */
class SingleKeyAuthWitnessProvider {
    constructor(privateKey, partialAddress) {
        this.privateKey = privateKey;
        this.partialAddress = partialAddress;
    }
    createAuthWitness(message) {
        const schnorr = new Schnorr();
        const signature = schnorr.constructSignature(message.toBuffer(), this.privateKey);
        const publicKey = generatePublicKey(this.privateKey);
        const witness = [...publicKey.toFields(), ...signature.toBuffer(), this.partialAddress];
        return Promise.resolve(new AuthWitness(message, witness));
    }
}
