// @aztec/circuit-types@0.28.1/dest/contract_data.js
import { FUNCTION_SELECTOR_NUM_BYTES, Fr, FunctionSelector, computeSaltedInitializationHash } from '@aztec/circuits.js';
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { randomBytes } from '@aztec/foundation/crypto';
import { EthAddress } from '@aztec/foundation/eth-address';
import { BufferReader, numToInt32BE, serializeArrayOfBufferableToVector, serializeToBuffer, } from '@aztec/foundation/serialize';
/**
 * Represents encoded contract function.
 */
export class EncodedContractFunction {
    constructor(
    /**
     * The function selector.
     */
    selector, 
    /**
     * Whether the function is internal.
     */
    isInternal, 
    /**
     * The function bytecode.
     */
    bytecode) {
        this.selector = selector;
        this.isInternal = isInternal;
        this.bytecode = bytecode;
    }
    /**
     * Serializes this instance into a buffer.
     * @returns Encoded buffer.
     */
    toBuffer() {
        const bytecodeBuf = Buffer.concat([numToInt32BE(this.bytecode.length), this.bytecode]);
        return serializeToBuffer(this.selector, this.isInternal, bytecodeBuf);
    }
    /**
     * Deserializes a contract function object from an encoded buffer.
     * @param buffer - The encoded buffer.
     * @returns The deserialized contract function.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const fnSelector = FunctionSelector.fromBuffer(reader.readBytes(FUNCTION_SELECTOR_NUM_BYTES));
        const isInternal = reader.readBoolean();
        return new EncodedContractFunction(fnSelector, isInternal, reader.readBuffer());
    }
    /**
     * Serializes this instance into a string.
     * @returns Encoded string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /**
     * Deserializes a contract function object from an encoded string.
     * @param data - The encoded string.
     * @returns The deserialized contract function.
     */
    static fromString(data) {
        return EncodedContractFunction.fromBuffer(Buffer.from(data, 'hex'));
    }
    /**
     * Creates a random contract function.
     * @returns A random contract function.
     */
    static random() {
        return new EncodedContractFunction(FunctionSelector.fromBuffer(randomBytes(4)), false, randomBytes(64));
    }
}
/**
 * A contract data blob, containing L1 and L2 addresses, public functions' bytecode, partial address and public key.
 * TODO(palla/purge-old-contract-deploy): Delete this class?
 */
export class ExtendedContractData {
    constructor(
    /** The base contract data: aztec & portal addresses. */
    contractData, 
    /** Artifacts of public functions. */
    publicFunctions, 
    /** Contract class id */
    contractClassId, 
    /** Salted init hash. */
    saltedInitializationHash, 
    /** Public key hash of the contract. */
    publicKeyHash) {
        this.contractData = contractData;
        this.publicFunctions = publicFunctions;
        this.contractClassId = contractClassId;
        this.saltedInitializationHash = saltedInitializationHash;
        this.publicKeyHash = publicKeyHash;
        this.bytecode = serializeArrayOfBufferableToVector(publicFunctions.map(fn => fn.toBuffer()));
    }
    /**
     * Gets the public function data or undefined.
     * @param selector - The function selector of the function to fetch.
     * @returns The public function data (if found).
     */
    getPublicFunction(selector) {
        return this.publicFunctions.find(fn => fn.selector.equals(selector));
    }
    /**
     * Serializes this instance into a buffer, using 20 bytes for the eth address.
     * @returns Encoded buffer.
     */
    toBuffer() {
        const contractDataBuf = this.contractData.toBuffer();
        return serializeToBuffer(contractDataBuf, this.bytecode, this.contractClassId, this.saltedInitializationHash, this.publicKeyHash);
    }
    /**
     * Serializes this instance into a string.
     * @returns Encoded string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /** True if this represents an empty instance. */
    isEmpty() {
        return ExtendedContractData.isEmpty(this);
    }
    /** True if the passed instance is empty . */
    static isEmpty(obj) {
        return (obj.contractData.isEmpty() &&
            obj.publicFunctions.length === 0 &&
            obj.contractClassId.isZero() &&
            obj.publicKeyHash.isZero() &&
            obj.saltedInitializationHash.isZero());
    }
    /**
     * Deserializes a contract data object from an encoded buffer, using 20 bytes for the eth address.
     * @param buffer - Byte array resulting from calling toBuffer.
     * @returns Deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const contractData = reader.readObject(ContractData);
        const publicFns = reader.readVector(EncodedContractFunction);
        const contractClassId = reader.readObject(Fr);
        const saltedInitializationHash = reader.readObject(Fr);
        const publicKeyHash = reader.readObject(Fr);
        return new ExtendedContractData(contractData, publicFns, contractClassId, saltedInitializationHash, publicKeyHash);
    }
    /**
     * Deserializes a contract data object from an encoded string, using 20 bytes for the eth address.
     * @param str - String resulting from calling toString.
     * @returns Deserialized instance.
     */
    static fromString(str) {
        return ExtendedContractData.fromBuffer(Buffer.from(str, 'hex'));
    }
    /**
     * Generate ContractData with random addresses.
     * @param contractData - Optional contract data to use.
     * @returns A random ExtendedContractData object.
     */
    static random(contractData) {
        return new ExtendedContractData(contractData ?? ContractData.random(), [EncodedContractFunction.random(), EncodedContractFunction.random()], Fr.random(), Fr.random(), Fr.random());
    }
    /** Generates empty extended contract data. */
    static empty() {
        return new ExtendedContractData(ContractData.empty(), [], Fr.ZERO, Fr.ZERO, Fr.ZERO);
    }
    /** Temporary method for creating extended contract data out of classes and instances */
    static fromClassAndInstance(contractClass, instance) {
        return new ExtendedContractData(new ContractData(instance.address, instance.portalContractAddress), contractClass.publicFunctions.map(f => new EncodedContractFunction(f.selector, f.isInternal, f.bytecode)), instance.contractClassId, computeSaltedInitializationHash(instance), instance.publicKeysHash);
    }
}
/**
 * A contract data blob, containing L1 and L2 addresses.
 * TODO(palla/purge-old-contract-deploy): Delete me
 */
export class ContractData {
    constructor(
    /**
     * The L2 address of the contract, as a field element (32 bytes).
     */
    contractAddress, 
    /**
     * The L1 address of the contract, (20 bytes).
     */
    portalContractAddress) {
        this.contractAddress = contractAddress;
        this.portalContractAddress = portalContractAddress;
    }
    /**
     * Serializes this instance into a buffer, using 20 bytes for the eth address.
     * @returns Encoded buffer.
     */
    toBuffer() {
        return serializeToBuffer(this.contractAddress, this.portalContractAddress);
    }
    /**
     * Serializes this instance into a string, using 20 bytes for the eth address.
     * @returns Encoded string.
     */
    toString() {
        return this.toBuffer().toString('hex');
    }
    /** True if all data is zero. */
    isEmpty() {
        return this.contractAddress.isZero() && this.portalContractAddress.isZero();
    }
    /**
     * Deserializes a contract data object from an encoded buffer, using 20 bytes for the eth address.
     * @param buffer - Byte array resulting from calling toBuffer.
     * @returns Deserialized instance.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const aztecAddr = AztecAddress.fromBuffer(reader);
        const ethAddr = new EthAddress(reader.readBytes(EthAddress.SIZE_IN_BYTES));
        return new ContractData(aztecAddr, ethAddr);
    }
    /**
     * Deserializes a contract data object from an encoded string, using 20 bytes for the eth address.
     * @param str - String resulting from calling toString.
     * @returns Deserialized instance.
     */
    static fromString(str) {
        return ContractData.fromBuffer(Buffer.from(str, 'hex'));
    }
    /**
     * Generate ContractData with random addresses.
     * @returns ContractData.
     */
    static random() {
        return new ContractData(AztecAddress.random(), EthAddress.random());
    }
    /** Generates an empty ContractData. */
    static empty() {
        return new ContractData(AztecAddress.ZERO, EthAddress.ZERO);
    }
}
