// @aztec/circuit-types@0.28.1/dest/l1_to_l2_message.js
import { AztecAddress } from '@aztec/foundation/aztec-address';
import { toBigIntBE, toBufferBE } from '@aztec/foundation/bigint-buffer';
import { randomInt, sha256 } from '@aztec/foundation/crypto';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
export class NewInboxLeaf {
    constructor(
    /** L2 block number in which the message will be included. */
    blockNumber, 
    /** Index of the leaf in L2 block message subtree. */
    index, 
    /** Leaf of the subtree. */
    leaf) {
        this.blockNumber = blockNumber;
        this.index = index;
        this.leaf = leaf;
    }
    toBuffer() {
        return serializeToBuffer([this.blockNumber, this.index, this.leaf]);
    }
    fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const blockNumber = toBigIntBE(reader.readBytes(32));
        const index = toBigIntBE(reader.readBytes(32));
        const leaf = reader.readObject(Fr);
        return new NewInboxLeaf(blockNumber, index, leaf);
    }
}
/**
 * L1AndL2Message and Index (in the merkle tree) as one type
 * TODO(#4492): Nuke the following when purging the old inbox
 */
export class L1ToL2MessageAndIndex {
    constructor(
    /** the index in the L1 to L2 Message tree. */
    index, 
    /** The message. */
    message) {
        this.index = index;
        this.message = message;
    }
    toBuffer() {
        return Buffer.concat([toBufferBE(this.index, 32), this.message.toBuffer()]);
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromString(data) {
        const buffer = Buffer.from(data, 'hex');
        return L1ToL2MessageAndIndex.fromBuffer(buffer);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const index = toBigIntBE(reader.readBytes(32));
        const message = L1ToL2Message.fromBuffer(reader);
        return new L1ToL2MessageAndIndex(index, message);
    }
}
/**
 * The format of an L1 to L2 Message.
 * TODO(#4492): Nuke the following when purging the old inbox
 */
export class L1ToL2Message {
    constructor(
    /**
     * The sender of the message on L1.
     */
    sender, 
    /**
     * The recipient of the message on L2.
     */
    recipient, 
    /**
     * The message content.
     */
    content, 
    /**
     * The hash of the spending secret.
     */
    secretHash, 
    /**
     * The deadline for the message.
     */
    deadline, 
    /**
     * The fee for the message.
     */
    fee, 
    /**
     * The entry key for the message - optional.
     */
    entryKey) {
        this.sender = sender;
        this.recipient = recipient;
        this.content = content;
        this.secretHash = secretHash;
        this.deadline = deadline;
        this.fee = fee;
        this.entryKey = entryKey;
    }
    /**
     * Returns each element within its own field so that it can be consumed by an acvm oracle call.
     * @returns The message as an array of fields (in order).
     */
    toFields() {
        return [
            ...this.sender.toFields(),
            ...this.recipient.toFields(),
            this.content,
            this.secretHash,
            new Fr(BigInt(this.deadline)),
            new Fr(BigInt(this.fee)),
        ];
    }
    toBuffer() {
        return serializeToBuffer(this.sender, this.recipient, this.content, this.secretHash, this.deadline, this.fee);
    }
    hash() {
        return Fr.fromBufferReduce(sha256(serializeToBuffer(...this.toFields())));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const sender = reader.readObject(L1Actor);
        const recipient = reader.readObject(L2Actor);
        const content = Fr.fromBuffer(reader);
        const secretHash = Fr.fromBuffer(reader);
        const deadline = reader.readNumber();
        const fee = reader.readNumber();
        return new L1ToL2Message(sender, recipient, content, secretHash, deadline, fee);
    }
    toString() {
        return this.toBuffer().toString('hex');
    }
    static fromString(data) {
        const buffer = Buffer.from(data, 'hex');
        return L1ToL2Message.fromBuffer(buffer);
    }
    static empty() {
        return new L1ToL2Message(L1Actor.empty(), L2Actor.empty(), Fr.ZERO, Fr.ZERO, 0, 0);
    }
    static random(entryKey) {
        return new L1ToL2Message(L1Actor.random(), L2Actor.random(), Fr.random(), Fr.random(), randomInt(1000), randomInt(1000), entryKey);
    }
}
/**
 * The sender of an L1 to L2 message.
 * TODO(#4492): Move to separate file when purging the old inbox
 */
export class L1Actor {
    constructor(
    /**
     * The sender of the message.
     */
    sender, 
    /**
     * The chain id on which the message was sent.
     */
    chainId) {
        this.sender = sender;
        this.chainId = chainId;
    }
    static empty() {
        return new L1Actor(EthAddress.ZERO, 0);
    }
    toFields() {
        return [this.sender.toField(), new Fr(BigInt(this.chainId))];
    }
    toBuffer() {
        return serializeToBuffer(this.sender, this.chainId);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const ethAddr = reader.readObject(EthAddress);
        const chainId = reader.readNumber();
        return new L1Actor(ethAddr, chainId);
    }
    static random() {
        return new L1Actor(EthAddress.random(), randomInt(1000));
    }
}
/**
 * The recipient of an L2 message.
 * TODO(#4492): Move to separate file when purging the old inbox
 */
export class L2Actor {
    constructor(
    /**
     * The recipient of the message.
     */
    recipient, 
    /**
     * The version of the protocol.
     */
    version) {
        this.recipient = recipient;
        this.version = version;
    }
    static empty() {
        return new L2Actor(AztecAddress.ZERO, 0);
    }
    toFields() {
        return [this.recipient.toField(), new Fr(BigInt(this.version))];
    }
    toBuffer() {
        return serializeToBuffer(this.recipient, this.version);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const aztecAddr = AztecAddress.fromBuffer(reader);
        const version = reader.readNumber();
        return new L2Actor(aztecAddr, version);
    }
    static random() {
        return new L2Actor(AztecAddress.random(), randomInt(1000));
    }
}
