// @aztec/circuit-types@0.28.1/dest/mocks.js
import { AztecAddress, CallRequest, Fr, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, MAX_REVERTIBLE_PUBLIC_CALL_STACK_LENGTH_PER_TX, Proof, } from '@aztec/circuits.js';
import { makeTuple } from '@aztec/foundation/array';
import { times } from '@aztec/foundation/collection';
import { randomBytes } from '@aztec/foundation/crypto';
import { SerializableContractInstance } from '@aztec/types/contracts';
import { FunctionL2Logs, Note, TxL2Logs } from './logs/index.js';
import { makePrivateKernelTailCircuitPublicInputs, makePublicCallRequest } from './mocks_to_purge.js';
import { ExtendedNote } from './notes/index.js';
import { Tx, TxHash } from './tx/index.js';
/**
 * Testing utility to create empty logs composed from a single empty log.
 */
export function makeEmptyLogs() {
    const functionLogs = [new FunctionL2Logs([Buffer.alloc(0)])];
    return new TxL2Logs(functionLogs);
}
export const randomTxHash = () => new TxHash(randomBytes(32));
export const mockTx = (seed = 1, logs = true) => {
    const tx = new Tx(makePrivateKernelTailCircuitPublicInputs(seed), new Proof(Buffer.alloc(0)), logs ? TxL2Logs.random(8, 3) : TxL2Logs.empty(), // 8 priv function invocations creating 3 encrypted logs each
    logs ? TxL2Logs.random(11, 2) : TxL2Logs.empty(), // 8 priv + 3 pub function invocations creating 2 unencrypted logs each
    times(MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, makePublicCallRequest));
    tx.data.endNonRevertibleData.publicCallStack = [
        tx.enqueuedPublicFunctionCalls[1].toCallRequest(),
        tx.enqueuedPublicFunctionCalls[0].toCallRequest(),
        CallRequest.empty(),
    ];
    tx.data.end.publicCallStack = makeTuple(MAX_REVERTIBLE_PUBLIC_CALL_STACK_LENGTH_PER_TX, i => tx.enqueuedPublicFunctionCalls[i + 2]?.toCallRequest() ?? CallRequest.empty()).reverse();
    return tx;
};
export const randomContractArtifact = () => ({
    name: randomBytes(4).toString('hex'),
    functions: [],
    events: [],
    fileMap: {},
});
export const randomContractInstanceWithAddress = () => SerializableContractInstance.random().withAddress(AztecAddress.random());
export const randomDeployedContract = () => ({
    artifact: randomContractArtifact(),
    instance: randomContractInstanceWithAddress(),
});
export const randomExtendedNote = ({ note = Note.random(), owner = AztecAddress.random(), contractAddress = AztecAddress.random(), txHash = randomTxHash(), storageSlot = Fr.random(), noteTypeId = Fr.random(), } = {}) => {
    return new ExtendedNote(note, owner, contractAddress, storageSlot, noteTypeId, txHash);
};
