// @aztec/circuits.js@0.28.1/dest/constants.gen.js
/* eslint-disable */
// GENERATED FILE - DO NOT EDIT, RUN yarn remake-constants
export const ARGS_LENGTH = 16;
export const RETURN_VALUES_LENGTH = 4;
export const MAX_NEW_NOTE_HASHES_PER_CALL = 16;
export const MAX_NEW_NULLIFIERS_PER_CALL = 16;
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL = 4;
export const MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL = 4;
export const MAX_NEW_L2_TO_L1_MSGS_PER_CALL = 2;
export const MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL = 16;
export const MAX_PUBLIC_DATA_READS_PER_CALL = 16;
export const MAX_NOTE_HASH_READ_REQUESTS_PER_CALL = 32;
export const MAX_NULLIFIER_READ_REQUESTS_PER_CALL = 2;
export const MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_CALL = 2;
export const MAX_NULLIFIER_KEY_VALIDATION_REQUESTS_PER_CALL = 1;
export const MAX_NEW_NOTE_HASHES_PER_TX = 64;
export const MAX_NON_REVERTIBLE_NOTE_HASHES_PER_TX = 8;
export const MAX_REVERTIBLE_NOTE_HASHES_PER_TX = 56;
export const MAX_NEW_NULLIFIERS_PER_TX = 64;
export const MAX_NON_REVERTIBLE_NULLIFIERS_PER_TX = 8;
export const MAX_REVERTIBLE_NULLIFIERS_PER_TX = 56;
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX = 8;
export const MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX = 8;
export const MAX_NON_REVERTIBLE_PUBLIC_CALL_STACK_LENGTH_PER_TX = 3;
export const MAX_REVERTIBLE_PUBLIC_CALL_STACK_LENGTH_PER_TX = 5;
export const MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 32;
export const MAX_NON_REVERTIBLE_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 16;
export const MAX_REVERTIBLE_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 16;
export const MAX_PUBLIC_DATA_READS_PER_TX = 32;
export const MAX_NON_REVERTIBLE_PUBLIC_DATA_READS_PER_TX = 16;
export const MAX_REVERTIBLE_PUBLIC_DATA_READS_PER_TX = 16;
export const MAX_NEW_L2_TO_L1_MSGS_PER_TX = 2;
export const MAX_NOTE_HASH_READ_REQUESTS_PER_TX = 128;
export const MAX_NULLIFIER_READ_REQUESTS_PER_TX = 8;
export const MAX_NULLIFIER_NON_EXISTENT_READ_REQUESTS_PER_TX = 8;
export const MAX_NULLIFIER_KEY_VALIDATION_REQUESTS_PER_TX = 4;
export const NUM_ENCRYPTED_LOGS_HASHES_PER_TX = 1;
export const NUM_UNENCRYPTED_LOGS_HASHES_PER_TX = 1;
export const NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP = 16;
export const VK_TREE_HEIGHT = 3;
export const FUNCTION_TREE_HEIGHT = 5;
export const NOTE_HASH_TREE_HEIGHT = 32;
export const PUBLIC_DATA_TREE_HEIGHT = 40;
export const NULLIFIER_TREE_HEIGHT = 20;
export const L1_TO_L2_MSG_TREE_HEIGHT = 16;
export const ROLLUP_VK_TREE_HEIGHT = 8;
export const ARTIFACT_FUNCTION_TREE_MAX_HEIGHT = 5;
export const NULLIFIER_TREE_ID = 0;
export const NOTE_HASH_TREE_ID = 1;
export const PUBLIC_DATA_TREE_ID = 2;
export const L1_TO_L2_MESSAGE_TREE_ID = 3;
export const ARCHIVE_TREE_ID = 4;
export const NOTE_HASH_SUBTREE_HEIGHT = 6;
export const NOTE_HASH_SUBTREE_SIBLING_PATH_LENGTH = 26;
export const NULLIFIER_SUBTREE_HEIGHT = 6;
export const PUBLIC_DATA_SUBTREE_HEIGHT = 5;
export const ARCHIVE_HEIGHT = 16;
export const NULLIFIER_SUBTREE_SIBLING_PATH_LENGTH = 14;
export const PUBLIC_DATA_SUBTREE_SIBLING_PATH_LENGTH = 35;
export const L1_TO_L2_MSG_SUBTREE_HEIGHT = 4;
export const L1_TO_L2_MSG_SUBTREE_SIBLING_PATH_LENGTH = 12;
export const FUNCTION_SELECTOR_NUM_BYTES = 4;
export const NUM_FIELDS_PER_SHA256 = 2;
export const ARGS_HASH_CHUNK_LENGTH = 32;
export const ARGS_HASH_CHUNK_COUNT = 32;
export const INITIALIZATION_SLOT_SEPARATOR = 1000000000;
export const INITIAL_L2_BLOCK_NUM = 1;
export const BLOB_SIZE_IN_BYTES = 126976;
export const MAX_PACKED_PUBLIC_BYTECODE_SIZE_IN_FIELDS = 15000;
export const MAX_PACKED_BYTECODE_SIZE_PER_PRIVATE_FUNCTION_IN_FIELDS = 500;
export const MAX_PACKED_BYTECODE_SIZE_PER_UNCONSTRAINED_FUNCTION_IN_FIELDS = 500;
export const REGISTERER_CONTRACT_CLASS_REGISTERED_MAGIC_VALUE = 0x6999d1e02b08a447a463563453cb36919c9dd7150336fc7c4d2b52f8n;
export const REGISTERER_PRIVATE_FUNCTION_BROADCASTED_MAGIC_VALUE = 0x1b70e95fde0b70adc30496b90a327af6a5e383e028e7a43211a07bcdn;
export const REGISTERER_UNCONSTRAINED_FUNCTION_BROADCASTED_MAGIC_VALUE = 0xe7af816635466f128568edb04c9fa024f6c87fb9010fdbffa68b3d99n;
export const DEPLOYER_CONTRACT_INSTANCE_DEPLOYED_MAGIC_VALUE = 0x85864497636cf755ae7bde03f267ce01a520981c21c3682aaf82a631n;
export const DEPLOYER_CONTRACT_ADDRESS = 0x0bffa876f07f9fe1802579dfef599810202f9c25b9a2f58921064a267d1ad1d3n;
export const L1_TO_L2_MESSAGE_ORACLE_CALL_LENGTH = 17;
export const MAX_NOTE_FIELDS_LENGTH = 20;
export const GET_NOTE_ORACLE_RETURN_LENGTH = 23;
export const MAX_NOTES_PER_PAGE = 10;
export const VIEW_NOTE_ORACLE_RETURN_LENGTH = 212;
export const AZTEC_ADDRESS_LENGTH = 1;
export const CALL_CONTEXT_LENGTH = 7;
export const CONTENT_COMMITMENT_LENGTH = 7;
export const CONTRACT_INSTANCE_LENGTH = 6;
export const CONTRACT_STORAGE_READ_LENGTH = 2;
export const CONTRACT_STORAGE_UPDATE_REQUEST_LENGTH = 2;
export const ETH_ADDRESS_LENGTH = 1;
export const FUNCTION_DATA_LENGTH = 4;
export const FUNCTION_LEAF_PREIMAGE_LENGTH = 5;
export const GLOBAL_VARIABLES_LENGTH = 6;
export const HEADER_LENGTH = 23;
export const L1_TO_L2_MESSAGE_LENGTH = 8;
export const L2_TO_L1_MESSAGE_LENGTH = 2;
export const NULLIFIER_KEY_VALIDATION_REQUEST_LENGTH = 4;
export const NULLIFIER_KEY_VALIDATION_REQUEST_CONTEXT_LENGTH = 5;
export const PARTIAL_STATE_REFERENCE_LENGTH = 6;
export const PRIVATE_CALL_STACK_ITEM_LENGTH = 215;
export const PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH = 210;
export const PUBLIC_CIRCUIT_PUBLIC_INPUTS_LENGTH = 202;
export const STATE_REFERENCE_LENGTH = 8;
export const TX_CONTEXT_DATA_LENGTH = 4;
export const TX_REQUEST_LENGTH = 10;
export const ENQUEUE_PUBLIC_FUNCTION_CALL_RETURN_LENGTH = 13;
export const GET_NOTES_ORACLE_RETURN_LENGTH = 674;
export const NOTE_HASHES_NUM_BYTES_PER_BASE_ROLLUP = 2048;
export const NULLIFIERS_NUM_BYTES_PER_BASE_ROLLUP = 2048;
export const PUBLIC_DATA_WRITES_NUM_BYTES_PER_BASE_ROLLUP = 2048;
export const CONTRACTS_NUM_BYTES_PER_BASE_ROLLUP = 32;
export const CONTRACT_DATA_NUM_BYTES_PER_BASE_ROLLUP = 64;
export const CONTRACT_DATA_NUM_BYTES_PER_BASE_ROLLUP_UNPADDED = 52;
export const L2_TO_L1_MSGS_NUM_BYTES_PER_BASE_ROLLUP = 64;
export const LOGS_HASHES_NUM_BYTES_PER_BASE_ROLLUP = 64;
export const NUM_MSGS_PER_BASE_PARITY = 4;
export const NUM_BASE_PARITY_PER_ROOT_PARITY = 4;
export var GeneratorIndex;
(function (GeneratorIndex) {
    GeneratorIndex[GeneratorIndex["NOTE_HASH"] = 1] = "NOTE_HASH";
    GeneratorIndex[GeneratorIndex["NOTE_HASH_NONCE"] = 2] = "NOTE_HASH_NONCE";
    GeneratorIndex[GeneratorIndex["UNIQUE_NOTE_HASH"] = 3] = "UNIQUE_NOTE_HASH";
    GeneratorIndex[GeneratorIndex["SILOED_NOTE_HASH"] = 4] = "SILOED_NOTE_HASH";
    GeneratorIndex[GeneratorIndex["NULLIFIER"] = 5] = "NULLIFIER";
    GeneratorIndex[GeneratorIndex["INITIALIZATION_NULLIFIER"] = 6] = "INITIALIZATION_NULLIFIER";
    GeneratorIndex[GeneratorIndex["OUTER_NULLIFIER"] = 7] = "OUTER_NULLIFIER";
    GeneratorIndex[GeneratorIndex["PUBLIC_DATA_READ"] = 8] = "PUBLIC_DATA_READ";
    GeneratorIndex[GeneratorIndex["PUBLIC_DATA_UPDATE_REQUEST"] = 9] = "PUBLIC_DATA_UPDATE_REQUEST";
    GeneratorIndex[GeneratorIndex["FUNCTION_DATA"] = 10] = "FUNCTION_DATA";
    GeneratorIndex[GeneratorIndex["FUNCTION_LEAF"] = 11] = "FUNCTION_LEAF";
    GeneratorIndex[GeneratorIndex["CONTRACT_DEPLOYMENT_DATA"] = 12] = "CONTRACT_DEPLOYMENT_DATA";
    GeneratorIndex[GeneratorIndex["CONSTRUCTOR"] = 13] = "CONSTRUCTOR";
    GeneratorIndex[GeneratorIndex["CONSTRUCTOR_ARGS"] = 14] = "CONSTRUCTOR_ARGS";
    GeneratorIndex[GeneratorIndex["CONTRACT_ADDRESS"] = 15] = "CONTRACT_ADDRESS";
    GeneratorIndex[GeneratorIndex["CONTRACT_LEAF"] = 16] = "CONTRACT_LEAF";
    GeneratorIndex[GeneratorIndex["CALL_CONTEXT"] = 17] = "CALL_CONTEXT";
    GeneratorIndex[GeneratorIndex["CALL_STACK_ITEM"] = 18] = "CALL_STACK_ITEM";
    GeneratorIndex[GeneratorIndex["CALL_STACK_ITEM_2"] = 19] = "CALL_STACK_ITEM_2";
    GeneratorIndex[GeneratorIndex["L1_TO_L2_MESSAGE_SECRET"] = 20] = "L1_TO_L2_MESSAGE_SECRET";
    GeneratorIndex[GeneratorIndex["L2_TO_L1_MSG"] = 21] = "L2_TO_L1_MSG";
    GeneratorIndex[GeneratorIndex["TX_CONTEXT"] = 22] = "TX_CONTEXT";
    GeneratorIndex[GeneratorIndex["PUBLIC_LEAF_INDEX"] = 23] = "PUBLIC_LEAF_INDEX";
    GeneratorIndex[GeneratorIndex["PUBLIC_DATA_LEAF"] = 24] = "PUBLIC_DATA_LEAF";
    GeneratorIndex[GeneratorIndex["SIGNED_TX_REQUEST"] = 25] = "SIGNED_TX_REQUEST";
    GeneratorIndex[GeneratorIndex["GLOBAL_VARIABLES"] = 26] = "GLOBAL_VARIABLES";
    GeneratorIndex[GeneratorIndex["PARTIAL_ADDRESS"] = 27] = "PARTIAL_ADDRESS";
    GeneratorIndex[GeneratorIndex["BLOCK_HASH"] = 28] = "BLOCK_HASH";
    GeneratorIndex[GeneratorIndex["SIDE_EFFECT"] = 29] = "SIDE_EFFECT";
    GeneratorIndex[GeneratorIndex["FEE_PAYLOAD"] = 30] = "FEE_PAYLOAD";
    GeneratorIndex[GeneratorIndex["TX_REQUEST"] = 33] = "TX_REQUEST";
    GeneratorIndex[GeneratorIndex["SIGNATURE_PAYLOAD"] = 34] = "SIGNATURE_PAYLOAD";
    GeneratorIndex[GeneratorIndex["VK"] = 41] = "VK";
    GeneratorIndex[GeneratorIndex["PRIVATE_CIRCUIT_PUBLIC_INPUTS"] = 42] = "PRIVATE_CIRCUIT_PUBLIC_INPUTS";
    GeneratorIndex[GeneratorIndex["PUBLIC_CIRCUIT_PUBLIC_INPUTS"] = 43] = "PUBLIC_CIRCUIT_PUBLIC_INPUTS";
    GeneratorIndex[GeneratorIndex["FUNCTION_ARGS"] = 44] = "FUNCTION_ARGS";
    GeneratorIndex[GeneratorIndex["AUTHWIT_INNER"] = 45] = "AUTHWIT_INNER";
    GeneratorIndex[GeneratorIndex["AUTHWIT_OUTER"] = 46] = "AUTHWIT_OUTER";
})(GeneratorIndex || (GeneratorIndex = {}));
