// @aztec/circuits.js@0.28.1/dest/structs/kernel/public_kernel_circuit_public_inputs.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { AggregationObject } from '../aggregation_object.js';
import { CombinedAccumulatedData, PublicAccumulatedNonRevertibleData, PublicAccumulatedRevertibleData, } from './combined_accumulated_data.js';
import { CombinedConstantData } from './combined_constant_data.js';
/**
 * Outputs from the public kernel circuits.
 * All Public kernels use this shape for outputs.
 */
export class PublicKernelCircuitPublicInputs {
    constructor(
    /**
     * Aggregated proof of all the previous kernel iterations.
     */
    aggregationObject, // Contains the aggregated proof of all previous kernel iterations
    /**
     * Accumulated side effects and enqueued calls that are not revertible.
     */
    endNonRevertibleData, 
    /**
     * Data accumulated from both public and private circuits.
     */
    end, 
    /**
     * Data which is not modified by the circuits.
     */
    constants, 
    /**
     * Indicates whether the setup kernel is needed.
     */
    needsSetup, 
    /**
     * Indicates whether the app logic kernel is needed.
     */
    needsAppLogic, 
    /**
     * Indicates whether the teardown kernel is needed.
     */
    needsTeardown, 
    /**
     * Indicates whether execution of the public circuit reverted.
     */
    reverted) {
        this.aggregationObject = aggregationObject;
        this.endNonRevertibleData = endNonRevertibleData;
        this.end = end;
        this.constants = constants;
        this.needsSetup = needsSetup;
        this.needsAppLogic = needsAppLogic;
        this.needsTeardown = needsTeardown;
        this.reverted = reverted;
        this.combined = undefined;
    }
    toBuffer() {
        return serializeToBuffer(this.aggregationObject, this.endNonRevertibleData, this.end, this.constants, this.needsSetup, this.needsAppLogic, this.needsTeardown, this.reverted);
    }
    get combinedData() {
        if (this.needsSetup || this.needsAppLogic || this.needsTeardown) {
            throw new Error('Cannot combine data when the circuit is not finished');
        }
        if (!this.combined) {
            this.combined = CombinedAccumulatedData.recombine(this.endNonRevertibleData, this.end, this.reverted);
        }
        return this.combined;
    }
    /**
     * Deserializes from a buffer or reader, corresponding to a write in cpp.
     * @param buffer - Buffer or reader to read from.
     * @returns A new instance of PublicKernelCircuitPublicInputs.
     */
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicKernelCircuitPublicInputs(reader.readObject(AggregationObject), reader.readObject(PublicAccumulatedNonRevertibleData), reader.readObject(PublicAccumulatedRevertibleData), reader.readObject(CombinedConstantData), reader.readBoolean(), reader.readBoolean(), reader.readBoolean(), reader.readBoolean());
    }
    static empty() {
        return new PublicKernelCircuitPublicInputs(AggregationObject.makeFake(), PublicAccumulatedNonRevertibleData.empty(), PublicAccumulatedRevertibleData.empty(), CombinedConstantData.empty(), false, false, false, false);
    }
    [inspect.custom]() {
        return `PublicKernelCircuitPublicInputs {
      aggregationObject: ${this.aggregationObject},
      endNonRevertibleData: ${inspect(this.endNonRevertibleData)},
      end: ${inspect(this.end)},
      constants: ${this.constants},
      needsSetup: ${this.needsSetup},
      needsAppLogic: ${this.needsAppLogic},
      needsTeardown: ${this.needsTeardown},
      reverted: ${this.reverted}
      }`;
    }
}
