// @aztec/aztec.js@0.28.1/dest/fee/native_fee_payment_method.js
var _a, _NativeFeePaymentMethod_GAS_TOKEN;
import { __classPrivateFieldGet } from "tslib";
import { FunctionData } from '@aztec/circuits.js';
import { FunctionSelector } from '@aztec/foundation/abi';
import { GasTokenAddress } from '@aztec/protocol-contracts/gas-token';
/**
 * Pay fee directly in the native gas token.
 */
export class NativeFeePaymentMethod {
    constructor() { }
    /**
     * Gets the native gas asset used to pay the fee.
     * @returns The asset used to pay the fee.
     */
    getAsset() {
        return __classPrivateFieldGet(_a, _a, "f", _NativeFeePaymentMethod_GAS_TOKEN);
    }
    /**
     * The contract responsible for fee payment. This will be the same as the asset.
     * @returns The contract address responsible for holding the fee payment.
     */
    getPaymentContract() {
        return __classPrivateFieldGet(_a, _a, "f", _NativeFeePaymentMethod_GAS_TOKEN);
    }
    /**
     * Fee payments in the native gas token are always public.
     * @returns false
     */
    isPrivateFeePayment() {
        return false;
    }
    /**
     * Creates a function call to pay the fee in gas token..
     * @param feeLimit - The maximum fee to be paid in gas token.
     * @returns A function call
     */
    getFunctionCalls(feeLimit) {
        return Promise.resolve([
            {
                to: __classPrivateFieldGet(_a, _a, "f", _NativeFeePaymentMethod_GAS_TOKEN),
                functionData: new FunctionData(FunctionSelector.fromSignature('check_balance(Field)'), false, false, false),
                args: [feeLimit],
            },
            {
                to: __classPrivateFieldGet(_a, _a, "f", _NativeFeePaymentMethod_GAS_TOKEN),
                functionData: new FunctionData(FunctionSelector.fromSignature('pay_fee(Field)'), false, false, false),
                args: [feeLimit],
            },
        ]);
    }
}
_a = NativeFeePaymentMethod;
_NativeFeePaymentMethod_GAS_TOKEN = { value: GasTokenAddress };
