// @aztec/aztec.js@0.28.1/dest/wallet/account_wallet.js
import { Fr } from '@aztec/circuits.js';
import { FunctionType } from '@aztec/foundation/abi';
import { ContractFunctionInteraction } from '../contract/contract_function_interaction.js';
import { BaseWallet } from './base_wallet.js';
/**
 * A wallet implementation that forwards authentication requests to a provided account.
 */
export class AccountWallet extends BaseWallet {
    constructor(pxe, account) {
        super(pxe);
        this.account = account;
    }
    createTxExecutionRequest(execs, fee) {
        return this.account.createTxExecutionRequest(execs, fee);
    }
    async createAuthWitness(message) {
        message = Buffer.isBuffer(message) ? Fr.fromBuffer(message) : message;
        const witness = await this.account.createAuthWitness(message);
        await this.pxe.addAuthWitness(witness);
        return witness;
    }
    /**
     * Returns a function interaction to set a message hash as authorized in this account.
     * Public calls can then consume this authorization.
     * @param message - Message hash to authorize.
     * @param authorized - True to authorize, false to revoke authorization.
     * @returns - A function interaction.
     */
    setPublicAuth(message, authorized) {
        if (authorized) {
            return new ContractFunctionInteraction(this, this.getAddress(), this.getApprovePublicAuthwitAbi(), [message]);
        }
        else {
            return this.cancelAuthWit(message);
        }
    }
    /**
     * Returns a function interaction to cancel a message hash as authorized in this account.
     * @param message - Message hash to authorize.
     * @returns - A function interaction.
     */
    cancelAuthWit(message) {
        const args = [message];
        return new ContractFunctionInteraction(this, this.getAddress(), this.getCancelAuthwitAbi(), args);
    }
    /** Returns the complete address of the account that implements this wallet. */
    getCompleteAddress() {
        return this.account.getCompleteAddress();
    }
    /** Returns the address of the account that implements this wallet. */
    getAddress() {
        return this.getCompleteAddress().address;
    }
    getApprovePublicAuthwitAbi() {
        return {
            name: 'approve_public_authwit',
            isInitializer: false,
            functionType: FunctionType.OPEN,
            isInternal: true,
            parameters: [
                {
                    name: 'message_hash',
                    type: { kind: 'field' },
                    visibility: 'private',
                },
            ],
            returnTypes: [],
        };
    }
    getCancelAuthwitAbi() {
        return {
            name: 'cancel_authwit',
            isInitializer: false,
            functionType: FunctionType.SECRET,
            isInternal: true,
            parameters: [
                {
                    name: 'message_hash',
                    type: { kind: 'field' },
                    visibility: 'private',
                },
            ],
            returnTypes: [],
        };
    }
}
