// @aztec/noir-contracts.js@0.28.1/src/AvmTest.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractInstanceWithAddress,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  FunctionSelectorLike,
  NoirCompiledContract,
  Point,
  PublicKey,
  Wallet,
  WrappedFieldLike,
  loadContractArtifact,
} from '@aztec/aztec.js';

import AvmTestContractArtifactJson from '../artifacts/avm_test_contract-AvmTest.json' assert { type: 'json' };

export const AvmTestContractArtifact = loadContractArtifact(AvmTestContractArtifactJson as NoirCompiledContract);

/**
 * Type-safe interface for contract AvmTest;
 */
export class AvmTestContract extends ContractBase {
  private constructor(instance: ContractInstanceWithAddress, wallet: Wallet) {
    super(instance, AvmTestContractArtifact, wallet);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, AvmTestContract.artifact, wallet) as Promise<AvmTestContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<AvmTestContract>(
      Point.ZERO,
      wallet,
      AvmTestContractArtifact,
      AvmTestContract.at,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<AvmTestContract>(
      publicKey,
      wallet,
      AvmTestContractArtifact,
      AvmTestContract.at,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified constructor method.
   */
  public static deployWithOpts<M extends keyof AvmTestContract['methods']>(
    opts: { publicKey?: PublicKey; method?: M; wallet: Wallet },
    ...args: Parameters<AvmTestContract['methods'][M]>
  ) {
    return new DeployMethod<AvmTestContract>(
      opts.publicKey ?? Point.ZERO,
      opts.wallet,
      AvmTestContractArtifact,
      AvmTestContract.at,
      Array.from(arguments).slice(1),
      opts.method ?? 'constructor',
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return AvmTestContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** avm_readStorageSingle() */
    avm_readStorageSingle: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getAddress() */
    avm_getAddress: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_nested_call_to_add(argA: field, argB: field) */
    avm_nested_call_to_add: ((argA: FieldLike, argB: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_getSender() */
    avm_getSender: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_addStorageMap(to: struct, amount: integer) */
    avm_addStorageMap: ((to: AztecAddressLike, amount: bigint | number) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_getTimestamp() */
    avm_getTimestamp: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_note_hash_exists(note_hash: field, leaf_index: field) */
    avm_note_hash_exists: ((note_hash: FieldLike, leaf_index: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_send_l2_to_l1_msg(recipient: struct, content: field) */
    avm_send_l2_to_l1_msg: ((recipient: EthAddressLike, content: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_new_nullifier(nullifier: field) */
    avm_new_nullifier: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getPortal() */
    avm_getPortal: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_setOpcodeSmallField() */
    avm_setOpcodeSmallField: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_setStorageMap(to: struct, amount: integer) */
    avm_setStorageMap: ((to: AztecAddressLike, amount: bigint | number) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_setReadStorageSingle(a: field) */
    avm_setReadStorageSingle: ((a: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_addArgsReturn(argA: field, argB: field) */
    avm_addArgsReturn: ((argA: FieldLike, argB: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_addU128(a: struct, b: struct) */
    avm_addU128: ((
      a: { lo: FieldLike; hi: FieldLike },
      b: { lo: FieldLike; hi: FieldLike },
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_keccak_hash(data: array) */
    avm_keccak_hash: ((data: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getFeePerL1Gas() */
    avm_getFeePerL1Gas: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getChainId() */
    avm_getChainId: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getVersion() */
    avm_getVersion: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_setOpcodeUint64() */
    avm_setOpcodeUint64: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_setOpcodeUint8() */
    avm_setOpcodeUint8: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_emit_unencrypted_log() */
    avm_emit_unencrypted_log: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getStorageAddress() */
    avm_getStorageAddress: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_sha256_hash(data: array) */
    avm_sha256_hash: ((data: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_emit_nullifier_and_check(nullifier: field) */
    avm_emit_nullifier_and_check: ((nullifier: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_poseidon_hash(data: array) */
    avm_poseidon_hash: ((data: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getFeePerDaGas() */
    avm_getFeePerDaGas: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getBlockNumber() */
    avm_getBlockNumber: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_getOrigin() */
    avm_getOrigin: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** compute_note_hash_and_nullifier(contract_address: struct, nonce: field, storage_slot: field, note_type_id: field, serialized_note: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: AztecAddressLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      note_type_id: FieldLike,
      serialized_note: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_readStorageList() */
    avm_readStorageList: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_raw_nested_call_to_add(argA: field, argB: field) */
    avm_raw_nested_call_to_add: ((argA: FieldLike, argB: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_setStorageList(a: field, b: field) */
    avm_setStorageList: ((a: FieldLike, b: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_raw_nested_static_call_to_add(argA: field, argB: field) */
    avm_raw_nested_static_call_to_add: ((argA: FieldLike, argB: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_setStorageSingle(a: field) */
    avm_setStorageSingle: ((a: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_nested_static_call_to_add(argA: field, argB: field) */
    avm_nested_static_call_to_add: ((argA: FieldLike, argB: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_getFeePerL2Gas() */
    avm_getFeePerL2Gas: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_l1_to_l2_msg_exists(msg_hash: field, msg_leaf_index: field) */
    avm_l1_to_l2_msg_exists: ((msg_hash: FieldLike, msg_leaf_index: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** avm_new_note_hash(note_hash: field) */
    avm_new_note_hash: ((note_hash: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_raw_nested_static_call_to_set_admin() */
    avm_raw_nested_static_call_to_set_admin: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_nullifier_collision(nullifier: field) */
    avm_nullifier_collision: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_setOpcodeUint32() */
    avm_setOpcodeUint32: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_pedersen_hash(data: array) */
    avm_pedersen_hash: ((data: FieldLike[]) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_nullifier_exists(nullifier: field) */
    avm_nullifier_exists: ((nullifier: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_readStorageMap(address: struct) */
    avm_readStorageMap: ((address: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** avm_nested_static_call_to_set_admin() */
    avm_nested_static_call_to_set_admin: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
